Chore: Addon documentation restructure (#13358)
storybookjs/storybook: docs/addons/writing-addons.md

---
title: 'Writing your own Storybook addon'
---

> One of Storybook's main features it's is robust addon ecosystem. Addons can be used to enhance your developer experience or workflows. We're going to guide you with the necessary steps to create your own first addon.

Reviewer: `is it's`

Author: Thanks for catching that one. Updated with the latest commit

</div>

Next create a `.babelrc.js` file in the root directory with the following:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/my-addon-babel-configuration.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

<div class="aside">
Because the addon we're about to write will use ES6 and rely on JSX, thus we'll need to include Babel and configure it with the proper presets and options. 
</div>

Change your `package.json` and add the following script to build the addon:

```json
{
  "scripts": {
    "build": "babel ./src --out-dir ./dist"
  },
}
```
<div class="aside">
Running <code>yarn build</code> at this stage will output the code into the <code>dist</code> directory, transpiled into a ES5 module ready to be installed into any Storybook. 
</div>

### Add a panel

Now let’s add a panel to Storybook. Create a new directory called `src` and inside a new file called `register.js`.

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-addon-panel-initial.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

> This is boilerplate code for any addon that adds a panel to Storybook, and there’s really not much going on here. In this case, we’re just adding a static `div` that renders when the panel is selected in Storybook’s UI.

Reviewer: I'm going to argue there is quite a bit happening here. We could say it's boilerplate. But, let's explain what exactly is happening. i.e., `register`, `add` and the `key` bit which is super important and without it you get errors.

Author: Updated with the latest commit.

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/my-addon-root-level-register.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

The sole purpose of making this change is to help with the registration process of the addon. By design Storybook looks for either a `preset.js` or a `register.js` file located at the root level. Omitting this file would require additional configuration from the addon consumer.

### Packaging and publishing

With the example above we showed how to create a bare-bones addon. Before publishing and distributing your own addon make sure you have met the following requirements:

- A `package.json file` with metadata about the addon.
- Peer dependencies of `react` and `@storybook/addons`.
- A `register.js` file at the root level written as an ES5 module.
- A `src` directory containing the ES6 addon code.
- A `dist` directory containing transpiled ES5 code on publish.
- A [GitHub](https://github.com/) account to host your code.
- A [NPM](https://www.npmjs.com/) account to publish the addon.

> For a good template of an addon that meets the requirements above, check out [@storybook/addon-controls](https://npmjs.com/package/@storybook/addon-controls).

Reviewer: I think we should find a standalone example. `@storybook/addon-controls` lives in the monorepo and relies on the monorepo's build system. It's really hard to figure out how the addon is actually built and published.

Author: @winkerVSbecks we could use the one you've built? WDYT? It's a standalone one outside the monorepo.

</div>

Next, create a `.babelrc.js` file in the root directory with the following:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/my-addon-babel-configuration.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

<div class="aside">
Because the addon we're about to write will use ES6 and rely on JSX, thus we'll need to include Babel and configure it with the proper presets and options. 
</div>

Change your `package.json` and add the following script to build the addon:

```json
{
  "scripts": {
    "build": "babel ./src --out-dir ./dist"
  }
}
```
<div class="aside">
Running <code>yarn build</code> at this stage will output the code into the <code>dist</code> directory, transpiled into a ES5 module ready to be installed into any Storybook. 
</div>

Finally, create a new directory called `src` and inside a new file called `preset.js` with the following:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/my-addon-preset-implementation.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

Presets are the way addons hook into Storybook. Among other tasks they allow you to:

- Add to [Storybook's UI](#add-a-panel).

> - Add to the [preview iframe](./writing-presets.md#preview-entries).

Reviewer 2: ```suggestion
- Add to the [preview iframe](./writing-presets.md#preview-entries)
```

> - Modify [babel](./writing-presets.md#babel) and [webpack settings](./writing-presets.md#webpack).

Reviewer 2: ```suggestion
- Modify [babel](./writing-presets.md#babel) and [webpack settings](./writing-presets.md#webpack)
```

For this example, we'll just modify Storybook's UI.

### Add a panel

Now let’s add a panel to Storybook. Inside the `src` directory, create a new file called `register.js` and add the following:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/my-addon-initial-panel-state.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

<div class="aside">
Make sure to include the <code>key</code> when you register the addon. This will prevent any issues when the addon renders.
</div> 

Going over the code snippet in more detail. When Storybook starts up:

> - It [registers](./addons-api.md#addonsregister) the addon.

Reviewer 2: Make this a numbered list and remove the periods

The [addon API](./addons-api.md) provides hooks like this so all of that communication can happen behind the scenes. That means you can focus on your addon's functionality.

### Using the addon with a story

When Storybook was initialized it provided a small set of examples stories. Change your `Button.stories.js` to the following:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'react/button-story-with-addon-example.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

After applying the changes to the story your Storybook UI will yield the following:

<video autoPlay muted playsInline loop>
  <source
    src="addon-final-stage-optimized.mp4"
    type="video/mp4"
  />
</video>

### Root level preset.js

Before publishing the addon, we'll need to make one final change. In the root directory of the addon create a new file called `preset.js` and add the following:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/my-addon-root-level-preset.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

> Making this change helps Storybook's addon registration. By design Storybook looks for either a `preset.js` or a `register.js` file located at the root level. Omitting this file would require additional configuration from the addon consumer.

Reviewer 2: ```suggestion
This auto-registers the addon without any additional configuration from the user. Storybook looks for either a `preset.js` or a `register.js` file located at the root level.
```

### Packaging and publishing

> With the example above we showed how to create a bare-bones addon. Before publishing and distributing your own addon make sure you have met the following requirements:

Reviewer 2: ```suggestion
Now that you've seen how to create a bare-bones addon, let's see how to share it with the community. Before we begin, make sure your addon meets the following requirements:
```

> - A `package.json file` with metadata about the addon.

Reviewer 2: - Remove periods (these aren't sentences)
- No need to repeat "A..." for every item

- Peer dependencies of `react` and `@storybook/addons`.
- A `preset.js` file at the root level written as an ES5 module.
- A `src` directory containing the ES6 addon code.
- A `dist` directory containing transpiled ES5 code on publish.
- A [GitHub](https://github.com/) account to host your code.
- A [NPM](https://www.npmjs.com/) account to publish the addon.

> For a good template of an addon that meets the requirements above, check out [storybook-addon-outline](https://www.npmjs.com/package/storybook-addon-outline).

Reviewer 2: ```suggestion
For example, check out [storybook-addon-outline](https://www.npmjs.com/package/storybook-addon-outline) to see a project that meets these requirements.
```

> ### Next steps

Reviewer 2: ```suggestion
### More guides and tutorials
```

In the previous example, we introduced the structure of an addon, but barely scratched the surface of what addons can do.

> To dive deeper we recommend [Learn Storybook’s “creating addons”](https://www.learnstorybook.com/intro-to-storybook/react/en/creating-addons/) tutorial. It’s an excellent walkthrough that covers the same ground as the above introduction, but goes further and leads you through the full process of creating a realistic addon. Or [how to build a Storybook addon](https://www.chromatic.com/blog/how-to-build-a-storybook-addon/) blog post that guides you through the steps of creating a standalone addon in great length.

Reviewer 2: ```suggestion
To dive deeper we recommend [Learn Storybook’s “creating addons”](https://www.learnstorybook.com/intro-to-storybook/react/en/creating-addons/) tutorial. It’s an excellent walkthrough that covers the same ground as the above introduction, but goes further and leads you through the full process of creating a realistic addon. 

[How to build a Storybook addon](https://www.chromatic.com/blog/how-to-build-a-storybook-addon/) shows you how to create a standalone addon in great detail.
```
